[CloudHill] iP

CloudHill opened the pull request:
### Sam

> Hello, I am Sam!

**Sam** is your _personal assistant_, helping you keep track of all your tasks!

She is:

* text-based ⌨️
* easy to learn
* ~~FAST~~ SUPER FAST to use 

All you need to do is:

1. Download
2. Double-click
3. Add your tasks
4. Let Sam manage your tasks for you

Features:

- [x]  Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

---

Sam was written using [Java](https://en.wikipedia.org/wiki/Java_(programming_language)), here is her `main` method:

```java
public static void main(String[] args) {
    new Sam("data", "sam.txt").run();
}
```

Commit:
Add Gradle support

CloudHill committed:
renamed Duke to Sam

CloudHill committed:
greet, echo, exit

CloudHill committed:
add characters

CloudHill committed:
use switch to process input

CloudHill committed:
moved Assets to its own file

CloudHill committed:
add and list items

CloudHill committed:
renamed MyList to TaskList

CloudHill committed:
add Task class

CloudHill committed:
use enum for command list

CloudHill committed:
support marking and unmarking tasks

CloudHill committed:
setup inheritance for task types

CloudHill committed:
split input into two

CloudHill committed:
changed task to title in task constructors

CloudHill committed:
support for todos, events, and deadlines

CloudHill committed:
text ui testing

CloudHill committed:
error handling

CloudHill committed:
use varargs for talk

CloudHill committed:
refactor main loop

CloudHill committed:
remove DEFAULT Command

CloudHill committed:
use exceptions to handle errors

CloudHill committed:
minor changes

CloudHill committed:
add support for deleting tasks

CloudHill committed:
refactor main method

CloudHill committed:
fix exception message

CloudHill committed:
add finally block

CloudHill committed:
update tests

CloudHill committed:
Support saving the task list

CloudHill committed:
Support loading the saved task list

CloudHill committed:
Store dates as LocalDate

CloudHill committed:
Merge branch 'branch-Level-7'

CloudHill committed:
Merge branch 'branch-Level-8'

CloudHill committed:
Make Sam more OOP

CloudHill committed:
Add Ui class

Assets class removed

CloudHill committed:
Add Storage class

CloudHill committed:
Add Parser class

CloudHill committed:
Move task id check to TaskList

CloudHill committed:
Factor out taskArgs from Sam

CloudHill committed:
Extract commands into Command classes

CloudHill committed:
Handle exception when parsing int

CloudHill committed:
Clean up unused imports

CloudHill committed:
Move exceptions into separate files

CloudHill committed:
Organize into packages

CloudHill committed:
Clean up exceptions thrown by commands

CloudHill committed:
Merge branch 'add-gradle-support'

CloudHill committed:
Update build.gradle

CloudHill committed:
Rename getCommand to ParseCommand

CloudHill committed:
Fix exception on empty add task args

CloudHill committed:
Add JUnit tests

CloudHill committed:
Add JavaDoc comments

CloudHill committed:
Comply with the coding standard

CloudHill committed:
Change generateList to getTasks to return a List of tasks

CloudHill committed:
Support finding tasks

CloudHill committed:
Merge branch 'branch-A-JavaDoc'

CloudHill committed:
Merge branch 'branch-A-CodingStandard'

CloudHill committed:
Merge branch 'branch-Level-9'

CloudHill committed:
Replace taskType enum with char

CloudHill committed:
Configure checkstyle

CloudHill committed:
Change Sam to accept a Path instead of strings

CloudHill committed:
Fix checkstyle errors

CloudHill committed:
Add GUI

CloudHill committed:
Merge branch 'branch-Level-10'

nus-jeff commented on src/main/java/sam/Ui.java:
```diff
@@ -0,0 +1,153 @@
+ * Handles user interaction.
+ */
+public class Ui {
+    private static final String LOGO =
+              " ██████╗ █████╗ ███╗   ███╗\n"
+            + "██╔════╝██╔══██╗████╗ ████║\n"
+            + "╚█████╗ ███████║██╔████╔██║\n"
+            + " ╚═══██╗██╔══██║██║╚██╔╝██║\n"
+            + "██████╔╝██║  ██║██║ ╚═╝ ██║\n"
+            + "╚═════╝ ╚═╝  ╚═╝╚═╝     ╚═╝";
+    private static final String USER =
+              " ███████\n"
+            + "████▀██▀█\n"
+            + "████▄██▄█\n"
+            + " ▀▀▀▀▀▀▀";
+    private static final String SAM =
+              "        ▄\n"
+            + " ▒▒██▒▒▓▓▀\n"
+            + "▒▒▀██▀▒▒▓▓\n"
+            + " █▄██▄███▓▓\n"
+            + "  ▀▀▀▀▀▀ ▓";
+
```
Really nice 👍

nus-jeff commented on src/main/java/sam/command/AddCommand.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage)
+            throws SamMissingTaskTitleException, SamMissingTaskValueException,
+            SamMissingTaskArgException, SamInvalidDateException, SamSaveFailedException,
+            SamUnknownCommandException {
```
Perhaps you would like to just throw its parent class to shorten the code?

nus-jeff commented on src/main/java/sam/command/AddCommand.java:
```diff
@@ -0,0 +1,78 @@
+        Task task = null;
+
+        switch (taskType) {
+        case 'T':
+            task = new ToDo(title);
+            break;
+        case 'E':
+            if (!taskArgs.containsKey("from") || !taskArgs.containsKey("to")) {
+                throw new SamMissingTaskArgException();
+            }
+            LocalDate from = Parser.parseDate(taskArgs.get("from"));
+            LocalDate to = Parser.parseDate(taskArgs.get("to"));
+            task = new Event(title, from, to);
+            break;
+        case 'D':
```
Maybe you would like to explicitly name what T, E and D are so readers can intuitively understand the code?

nus-jeff commented on src/main/java/sam/command/FindCommand.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        List<String> list = tasks.findTasks(args);
```
Just a minor suggestion to use lists as the variable name to represent collections instead?

nus-jeff reviewed (commented):
Code Quality looks really good. My suggestion would probably be only to consider grouping the Exceptions together?

CloudHill committed:
Remove text ui test

CloudHill committed:
Refactor exit code

CloudHill committed:
Use isEmpty in FindCommand

CloudHill committed:
Add assertions

Java assertions aren't used anywhere in the code.

Java assertions help document and verify important assumptions
that should hold at various points in the code.

Let's add Java assertions to places in the code where some assumption
about the program state is made.

For the most part, assertions are added
where null values should be impossible

CloudHill committed:
Fix checkstyle errors

CloudHill committed:
Refactor Sam's dialogs into an enum

Sam's dialogs are scattered everywhere as String literals.

Having the dialogs in a centralized position will make them
easier to keep track of.

Let's add a Dialog enum under Ui where all of Sam's dialogs
will be defined.

The Dialog enum is put under UI since dialogs
are related to the UI (as feedback for the user)

CloudHill committed:
Use constants for date patterns

CloudHill committed:
Refactor magic numbers in Ui

Ui uses some magic numbers to define the dimensions of UI components.

Magic numbers should be avoided as their meaning can be confusing.

Let's declare named variables for these numbers.

CloudHill committed:
Refactor happy paths in TaskList

CloudHill committed:
Rename variable in TaskList for consistency

CloudHill committed:
Hide fields in Task

CloudHill committed:
Merge pull request #1 from CloudHill/branch-A-Assertions

Add assertions

CloudHill committed:
Merge pull request #2 from CloudHill/branch-A-CodeQuality

Refactor code

CloudHill committed:
Fix checkstyle error

CloudHill committed:
Set up CI

CloudHill committed:
Fix wrong dialogs

CloudHill committed:
Support editing tasks

Syntax: `edit <id> /key1 value1 /key2 value2 ...`
Use the key `/title` to update the title of the task

CloudHill committed:
Support cloning tasks

Syntax: `clone <id> /key1 value1 /key2 value2 ...`
Keys and values are optional and can be
used to update details of the cloned task
Use the key `/title` to update the title of the task

CloudHill committed:
Change setTaskDone to return the updated Task

CloudHill committed:
Merge branch 'branch-C-Update'

CloudHill committed:
Move the Dialog enum to its own file

CloudHill committed:
Create new package ui

CloudHill committed:
Apply SLAP to setStage

CloudHill committed:
Fix long method in AddCommand

CloudHill committed:
Render tasks as a UI component

CloudHill committed:
Use fxml for the MainWindow

CloudHill committed:
Fix task index

CloudHill committed:
Use images for user and sam

CloudHill committed:
Improve task card

CloudHill committed:
Use fxml for DialogBox

+ some TaskCard changes

CloudHill committed:
Update parser unit tests

CloudHill committed:
Add icon and Ui.png

CloudHill committed:
Fix checkStyle issues

CloudHill committed:
Update main class name

CloudHill committed:
Fix find command count

CloudHill committed:
Update sam's avatar

CloudHill committed:
Update user guide
